internal/resolve/catalog_test.go

```diff
@@ -963,3 +963,150 @@ func TestMultipleChannels(t *testing.T) {
 	assert.Equal(t, bsemver.MustParse("2.0.0"), *gotVersion)
 	assert.Equal(t, ptr.To(packageDeprecation(pkgName)), gotDeprecation)
 }
+
+func TestAllCatalogsDisabled(t *testing.T) {
+	pkgName := randPkg()
+	listCatalogs := func(ctx context.Context, options ...client.ListOption) ([]catalogd.ClusterCatalog, error) {
+		return []catalogd.ClusterCatalog{
+			{
+				Spec: catalogd.ClusterCatalogSpec{
+					Availability: "Disabled",
+				},
+			},
+			{
+				Spec: catalogd.ClusterCatalogSpec{
+					Availability: "Disabled",
+				},
+			},
+		}, nil
+	}
+
+	getPackage := func(ctx context.Context, cat *catalogd.ClusterCatalog, packageName string) (*declcfg.DeclarativeConfig, error) {
+		return genPackage(pkgName), nil
+	}
+
+	r := CatalogResolver{
+		WalkCatalogsFunc: CatalogWalker(listCatalogs, getPackage),
+	}
+
+	ce := buildFooClusterExtension(pkgName, []string{}, ">=1.0.0", ocv1alpha1.UpgradeConstraintPolicyCatalogProvided)
+	_, _, _, err := r.Resolve(context.Background(), ce, nil)
+	require.Error(t, err)
+	assert.Contains(t, err.Error(), "no enabled catalogs found for package")
+}
+
+func TestSomeCatalogsDisabled(t *testing.T) {
+	pkgName := randPkg()
+	listCatalogs := func(ctx context.Context, options ...client.ListOption) ([]catalogd.ClusterCatalog, error) {
+		return []catalogd.ClusterCatalog{
+			{
+				Spec: catalogd.ClusterCatalogSpec{
+					Availability: "Enabled",
+				},
+			},
+			{
+				Spec: catalogd.ClusterCatalogSpec{
+					Availability: "Disabled",
+				},
+			},
+		}, nil
+	}
+
+	getPackage := func(ctx context.Context, cat *catalogd.ClusterCatalog, packageName string) (*declcfg.DeclarativeConfig, error) {
+		return genPackage(pkgName), nil
+	}
+
+	r := CatalogResolver{
+		WalkCatalogsFunc: CatalogWalker(listCatalogs, getPackage),
+	}
+
+	ce := buildFooClusterExtension(pkgName, []string{}, ">=1.0.0", ocv1alpha1.UpgradeConstraintPolicyCatalogProvided)
+	gotBundle, gotVersion, _, err := r.Resolve(context.Background(), ce, nil)
+	require.NoError(t, err)
+	require.NotNil(t, gotBundle)
+	require.Equal(t, bsemver.MustParse("3.0.0"), *gotVersion)
+}
+
+func TestPriorityRespectedWithDisabledCatalogs(t *testing.T) {
+	pkgName := randPkg()
+	listCatalogs := func(ctx context.Context, options ...client.ListOption) ([]catalogd.ClusterCatalog, error) {
+		return []catalogd.ClusterCatalog{
+			{
+				Spec: catalogd.ClusterCatalogSpec{
+					Priority:     1,
+					Availability: "Enabled",
+				},
+			},
+			{
+				Spec: catalogd.ClusterCatalogSpec{
+					Priority:     0,
+					Availability: "Disabled",
+				},
+			},
+		}, nil
+	}
+
+	getPackage := func(ctx context.Context, cat *catalogd.ClusterCatalog, packageName string) (*declcfg.DeclarativeConfig, error) {
+		return genPackage(pkgName), nil
+	}
+
+	r := CatalogResolver{
+		WalkCatalogsFunc: CatalogWalker(listCatalogs, getPackage),
+	}
```

Comment:
It seems to me like if we managed to break the fact that disabled catalogs are ignored, this test would still pass?
1. The catalog with higher priority is enabled, so we'd resolve from it anyway.
2. If the disabled catalog was higher priority (and we have broken logic for ignoring disabled catalogs), `getPackage` would be called for both catalogs, which means that the resolver would still find the same bundle.
3. If we also broke the logic for considering enabled catalogs, this test looks like it would _still_ pass because `getPackage` would be called for the disabled catalog and it would still return the same result that the assertions expect.